### PR TITLE
feat: Support Github App mode

### PR DESCRIPTION
**Pull request by armenzg**
This adds support for running the app as a Github App.

If fetching the Github App private key from Google Secret Manager fails, we will fallback to using the Github Personal Access Token.

In the future, we will only support PAT usage for local development since creating a Github App for local development is a bit overkill.

### CODE REVIEW

File: .envrc [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# shellcheck shell=bash
```
Comment (armenzg) [inline .envrc:1]: This file could go in a separate PR.

File: .gitignore [hunk 2]
```diff
@@ -130,3 +130,6 @@ dmypy.json
 
 # Ignore generated credentials from google-github-actions/auth
 gha-creds-*.json
+
+# Ignore PEM
+*.pem
```
Comment (armenzg) [inline .gitignore:135]: Useful when testing the Github App set up locally.

File: README.md [hunk 3]
```diff
@@ -6,10 +6,21 @@ This app allows your organization to trace Github Actions with Sentry. You can u
 It works by listening to a Github workflow events via a webhook in your repository. These events are the stored in Sentry as performance transactions.
 
 `github_sdk.py` has the generic logic to submit Github jobs as transactions. Eventually this file could be released separatedly.
-`handle_event.py`: Business logic goes here.
+`event_handler.py`: Business logic goes here.
```
Comment (armenzg) [inline README.md:9]: It was renamed in another PR.

File: requirements.txt [hunk 4]
```diff
@@ -1,3 +1,4 @@
-flask>=2.0.3
-sentry-sdk[flask]>=1.5.8
-requests>=2.27.1
+PyJWT[crypto]==2.4.0
```
Comment (armenzg) [inline requirements.txt:1]: The optional crypto package is needed since we're dealing with RSA algorithms.

File: src/main.py [hunk 5]
```diff
@@ -26,15 +27,27 @@
...
+    # You can create an .env file and place the token in it
+    TOKEN = os.environ["GH_TOKEN"]
+
 # Where to report Github actions transactions
 SENTRY_GITHUB_DSN = os.environ.get("SENTRY_GITHUB_DSN")
 
 app = Flask(__name__)
-handler = EventHandler(token=GH_TOKEN, dsn=SENTRY_GITHUB_DSN)
+handler = EventHandler(
+    secret=os.environ.get("GH_WEBHOOK_SECRET"),
```
Comment (armenzg) [inline src/main.py:47]: It seems I forgot to add this in the previous PR, thus, we're currently not checking the secret.

**Comment by armenzg** [comment]
@asottile-sentry Hey, could you have a quick look? I'm more interested on getting feedback for the `github_app.py` module.

I've got this working for the CLI. Next, I'm going to make it work for my local dev app by sending real events from the `armenzg-dev` org and then add Google Secrets support.

```
python cli.py https://github.com/armenzg-dev/github-actions-app/runs/6613515416
```

File: cli.py [hunk 6]
```diff
@@ -1,25 +1,45 @@
...
 from src.github_sdk import GithubClient
+from src.github_app import get_access_tokens
+
 
 # Point this script to the URL of a job and we will trace it
 # You give us this https://github.com/getsentry/sentry/runs/5759197422?check_suite_focus=true
 # Or give it a path to a file with a webhook payload
 # e.g. tests/fixtures/jobA/job.json
 if __name__ == "__main__":
     argument = sys.argv[1]
```
Comment (asottile-sentry) [inline cli.py:16]: personally I like to keep this as short as possible as it's untestable otherwise

my commandlines always end with this:

```python
def main() -> int:
    ...

if __name__ == '__main__':
    raise SystemExit(main())
```

https://youtu.be/sv46294LoP8

File: src/github_app.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    return private_key
+
+
+def get_jwt_token(private_key, app_id):
+    payload = {
+        # issued at time, 60 seconds in the past to allow for clock drift
+        "iat": dt.utcnow() - timedelta(minutes=1),
+        # JWT expiration time (10 minute maximum)
+        "exp": dt.utcnow() + timedelta(minutes=10),
+        # GitHub App's identifier
```
Comment (asottile-sentry) [inline src/github_app.py:28]: interesting note about clock drift -- for my app I use this and it works:

```python
'iat': int(time.time()),
'exp': int(time.time()) + 300,
```

File: src/github_app.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    headers = {
+        "Accept": "application/vnd.github.v3+json",
+        "Authorization": f"Bearer {token}",
+    }
+    req = requests.get("https://api.github.com/app", headers=headers)
+    req.raise_for_status()
+    response = req.json()
+
+    # The second value tracks the org this token is useful for
+    return response["owner"]["login"]
```
Comment (asottile-sentry) [inline src/github_app.py:44]: this will get the owner of the *app*, I'm not sure if that's what's desired -- usually you want to get the installation by id or by org

this will coincidentally return the right thing if the app is created in the org and only used in the org however

File: src/github_app.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    installations = req.json()
+
+    # From docs: Installation access tokens have the permissions
+    # configured by the GitHub App and expire after one hour.
+    for inst in installations:
+        req = requests.post(url=inst["access_tokens_url"], headers=headers)
+        req.raise_for_status()
+        access_tokens[inst["account"]["login"]] = req.json()
+
+    return access_tokens
```
Comment (asottile-sentry) [inline src/github_app.py:67]: typically the webhook you'll get back with have the specific installation (installations are 1:1(user/org)) and you'll use that install id with the jwt to get the token directly

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -5,8 +5,8 @@ on:
     paths-ignore:
       - "**.md"
   push:
-    branches:
-      - "main"
+    # branches:
+    #   - "main"
```
Comment (armenzg) [inline .github/workflows/ci.yml:9]: This will be undone before merging.

File: src/github_app.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    return {
+        "Accept": "application/vnd.github.v3+json",
+        "Authorization": f"Bearer {jwt_token}",
+    }
+
+
+# From docs: Installation access tokens have the permissions
+# configured by the GitHub App and expire after one hour.
+# XXX: Create solution to get a new token
+def get_token(installation_id, headers):
```
Comment (armenzg) [inline src/github_app.py:45]: This PR is not complete. I will be adding the ability to refresh the token.

File: src/web_app_handler.py [hunk 12]
```diff
@@ -42,13 +46,20 @@ def valid_signature(self, body, headers):
             signature = headers["X-Hub-Signature-256"].replace("sha256=", "")
             body_signature = hmac.new(
                 self.config.gh.webhook_secret.encode(),
-                msg=str(body).encode(),
+                msg=body,
```
Comment (armenzg) [inline src/web_app_handler.py:49]: I honestly thought this got fixed in the PR where we fixed the GH webhook secret.
Anyway, I will do a deployment tomorrow with what is on `main` and revert if it's broken.

File: src/web_app_handler.py [hunk 13]
```diff
@@ -42,13 +46,20 @@ def valid_signature(self, body, headers):
...
                 digestmod="sha256",
             ).hexdigest()
             return hmac.compare_digest(body_signature, signature)
 
 
+class GithubAppConfig(NamedTuple):
```
Comment (armenzg) [inline src/web_app_handler.py:55]: I'm loving `NamedTuple`. Thanks!

File: tests/test_web_app_handler.py [hunk 14]
```diff
@@ -109,9 +110,13 @@ def test_invalid_signature(monkeypatch, webhook_event):
 def test_handle_event_with_secret(monkeypatch, webhook_event):
     monkeypatch.setenv("GH_WEBHOOK_SECRET", "fake_secret")
     handler = WebAppHandler(dry_run=True)
+    handler.gh_client.send_trace = MagicMock()
     reason, http_code = handler.handle_event(
         data=webhook_event["payload"],
         headers=webhook_event["headers"],
     )
+    handler.gh_client.send_trace.assert_called_with(
```
Comment (armenzg) [inline tests/test_web_app_handler.py:118]: I don't know how this test passed in the past. It definitely should have been needing mocking responses since we should have been hitting Github's API.

**Review by armenzg — COMMENTED** [review]
This PR is not complete as I need to refresh the tokens within the hour.

File: src/github_app.py (hunk 11 above)
Comment (asottile-sentry) [reply src/github_app.py:45]: I wouldn't bother with refresh -- I usually grab a new token for every webhook (same number of requests as trying to refresh it)

there is a token deletion though that's usually a best practice at the end of the webhook (I've got a context manager for mine!) -- `DELETE https://api.github.com/installation/token` (it'll delete the authorized token)

File: tests/test_web_app_handler.py [hunk 15]
```diff
@@ -109,9 +110,13 @@ def test_invalid_signature(monkeypatch, webhook_event):
 def test_handle_event_with_secret(monkeypatch, webhook_event):
     monkeypatch.setenv("GH_WEBHOOK_SECRET", "fake_secret")
     handler = WebAppHandler(dry_run=True)
+    handler.gh_client.send_trace = MagicMock()
```
Comment (asottile-sentry) [inline tests/test_web_app_handler.py:113]: I would use `with mock.patch.object(handler.gh_client, 'send_trace') as mck:` that way it is undone (this'll also appease mypy when we add it since mypy forbids assigning to methods)

**Review by asottile-sentry — APPROVED** [review]
otherwise looks good!

File: README.md [hunk 16]
```diff
@@ -6,10 +6,43 @@ This app allows your organization to trace Github Actions with Sentry. You can u
 It works by listening to a Github workflow events via a webhook in your repository. These events are the stored in Sentry as performance transactions.
 
 `github_sdk.py` has the generic logic to submit Github jobs as transactions. Eventually this file could be released separatedly.
-`handle_event.py`: Business logic goes here.
+`web_app_handler.py`: Web app logic goes here.
 `main.py` contains the code to respond to webhook events.
 
-## Set up
+You can set up this app following the instructions under "Self-hosted". We also have a Github app to increase the security, however, it can only be installed in the `getsentry` Github org. In the future, we may make it pubclicly available. There's few things we need to figure out before we do.
```
Comment (armenzg) [inline README.md:12]: For now, it can only be installed in the org `armenzg-dev`. Let's see if we figure out transferring correctly.

File: src/github_app.py [hunk 17]
```diff
@@ -1,53 +1,50 @@
...
+        resp = req.json()
+        # This token expires in an hour
+        return resp["token"]
+
+    # This is executed when we enter a with block
+    def __enter__(self):
+        return self.get_token()
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        requests.delete("https://api.github.com/installation/token")
```
Comment (asottile-sentry) [inline src/github_app.py:32]: these are really difficult to get correct imo -- it's much easier to use `@contextlib.contextmanager`: https://youtu.be/ucGpcA9r4hU

File: README.md [hunk 18]
```diff
@@ -6,10 +6,43 @@ This app allows your organization to trace Github Actions with Sentry. You can u
...
+
+## Github app
+
+Currently, it is only used internally. We do not have a way of mapping CI events from one org to a specific Sentry dsn.
```
Comment (armenzg) [inline README.md:16]: In the future, external orgs that would install the app _could_ define their org DSN in a Github secret that we would read.

File: tests/test_web_app_handler.py [hunk 19]
```diff
@@ -50,7 +51,8 @@ def test_not_completed_workflow():
     assert http_code == 200
 
 
-def test_missing_workflow_job():
+def test_missing_workflow_job(monkeypatch):
+    monkeypatch.delenv("GH_APP_ID", raising=False)
```
Comment (armenzg) [inline tests/test_web_app_handler.py:55]: When I do local development, I sometimes use `pytest` directly, thus, it will read my `.env` variables and think that I'm executing within the context of a Github app.

File: src/github_app.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
```
Comment (asottile-sentry) [inline src/github_app.py:32]: here's what I mean -- also I fixed the token deletion since it was missing the auth

```suggestion
    # From docs: Installation access tokens have the permissions
    # configured by the GitHub App and expire after one hour.
    @contextlib.contextmanager
    def get_token(self) -> Generator[str, None, None]:
        req = requests.post(
            url=f"https://api.github.com/app/installations/{self.installation_id}/access_tokens",
            headers=self.headers,
        )
        req.raise_for_status()
        resp = req.json()
        try:
            # This token expires in an hour
            yield resp["token"]
        finally:
            requests.delete("https://api.github.com/installation/token", headers={"Authorization": f"token {resp['token']}"})
```

**Review by asottile-sentry — APPROVED** [review]
